File: src/main/java/com/quantcapital/entities/Bar.java
package com.quantcapital.entities;

import com.fasterxml.jackson.annotation.JsonFormat;
import com.quantcapital.entities.constant.Frequency;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.math.BigDecimal;
import java.math.RoundingMode;
import java.time.LocalDateTime;

/**
 * K线数据实体
 * 
 * 包含OHLC价格数据、成交量以及各种技术指标。
 * 支持多个频率的K线数据，是量化分析的基础数据结构。
 * 
 * @author QuantCapital Team
 */
@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class Bar {
    
    /** 标的代码 */
    private String symbol;
    
    /** 时间戳 */
    @JsonFormat(pattern = "yyyy-MM-dd HH:mm:ss")
    private LocalDateTime datetime;
    
    /** 数据频率 */
    private Frequency frequency;
    
    // ==================== 基础OHLC数据 ====================
    
    /** 开盘价 */
    private double open;
    
    /** 最高价 */
    private double high;
    
    /** 最低价 */
    private double low;
    
    /** 收盘价 */
    private double close;
    
    /** 成交量（股） */
    private long volume;
    
    /** 成交额（元） */
    private double amount;
    
    /** 换手率（%） */
    private Double turnover;
    
    // ==================== 技术指标 ====================
    
    /** 5日移动平均线 */
    private Double ma5;
    
    /** 20日移动平均线 */
    private Double ma20;
    
    /** 60日移动平均线 */
    private Double ma60;
    
    /** MACD DIF线 */
    private Double macdDif;
    
    /** MACD DEA线 */
    private Double macdDea;
    
    /** MACD柱状图 */
    private Double macdHistogram;
    
    /** RSI相对强弱指标(0-100) */
    private Double rsi14;
    
    /** 布林带上轨 */
    private Double bollUpper;
    
    /** 布林带下轨 */
    private Double bollLower;
    
    // ==================== 基本面数据 ====================
    
    /** 总市值（亿元） */
    private Double marketCap;
    
    /** 流通市值（亿元） */
    private Double circulatingMarketCap;
    
    /** 是否ST股票 */
    private boolean isSt;
    
    /** 是否新股/次新股 */
    private boolean isNewStock;
    
    // ==================== 计算方法 ====================
    
    /**
     * 计算价格变化幅度
     * 
     * @return 价格变化幅度 (close - open) / open
     */
    public double getPriceChange() {
        if (open <= 0) {
            return 0.0;
        }
        return (close - open) / open;
    }
    
    /**
     * 计算价格变化幅度（百分比）
     * 
     * @return 价格变化百分比，保留2位小数
     */
    public double getPriceChangePercent() {
        return BigDecimal.valueOf(getPriceChange() * 100)
                .setScale(2, RoundingMode.HALF_UP)
                .doubleValue();
    }
    
    /**
     * 计算振幅
     * 
     * @return 振幅 (high - low) / open
     */
    public double getAmplitude() {
        if (open <= 0) {
            return 0.0;
        }
        return (high - low) / open;
    }
    
    /**
     * 计算上影线长度比例
     * 
     * @return 上影线长度占总振幅的比例
     */
    public double getUpperShadowRatio() {
        if (high <= low) {
            return 0.0;
        }
        double bodyHigh = Math.max(open, close);
        return (high - bodyHigh) / (high - low);
    }
    
    /**
     * 计算下影线长度比例
     * 
     * @return 下影线长度占总振幅的比例
     */
    public double getLowerShadowRatio() {
        if (high <= low) {
            return 0.0;
        }
        double bodyLow = Math.min(open, close);
        return (bodyLow - low) / (high - low);
    }
    
    /**
     * 判断是否为阳线
     * 
     * @return 是否为阳线
     */
    public boolean isGreen() {
        return close > open;
    }
    
    /**
     * 判断是否为阴线
     * 
     * @return 是否为阴线
     */
    public boolean isRed() {
        return close < open;
    }
    
    /**
     * 判断是否为十字星（开盘价和收盘价几乎相等）
     * 
     * @return 是否为十字星
     */
    public boolean isDoji() {
        if (open <= 0) {
            return false;
        }
        return Math.abs((close - open) / open) < 0.001; // 0.1%的阈值
    }
    
    /**
     * 判断是否有成交量异常放大
     * 
     * @param avgVolume 平均成交量
     * @param threshold 放大倍数阈值
     * @return 是否成交量异常
     */
    public boolean isVolumeAbnormal(long avgVolume, double threshold) {
        if (avgVolume <= 0) {
            return false;
        }
        return volume > avgVolume * threshold;
    }
    
    /**
     * 获取典型价格（HLC平均价）
     * 
     * @return 典型价格
     */
    public double getTypicalPrice() {
        return (high + low + close) / 3.0;
    }
    
    /**
     * 获取加权平均价格（VWAP近似）
     * 
     * @return 加权平均价格
     */
    public double getWeightedPrice() {
        return (open + high + low + close) / 4.0;
    }
    
    /**
     * 数据校验
     * 
     * @return 数据是否有效
     */
    public boolean isValid() {
        return symbol != null && !symbol.trim().isEmpty()
                && datetime != null
                && open > 0 && high > 0 && low > 0 && close > 0
                && high >= Math.max(open, close)
                && low <= Math.min(open, close)
                && volume >= 0
                && amount >= 0;
    }
    
    @Override
    public String toString() {
        return String.format("Bar[%s %s O:%.2f H:%.2f L:%.2f C:%.2f V:%d A:%.0f]",
                symbol, datetime, open, high, low, close, volume, amount);
    }
}

File: quantcapital/entities/bar.py
"""
K线数据实体定义

包含OHLC基础数据、技术指标和基本面数据的完整K线数据结构。
"""

from datetime import datetime
from dataclasses import dataclass
from typing import Optional
from enum import Enum


class Frequency(Enum):
    """K线频率枚举"""
    HOURLY = "H"     # 小时线
    DAILY = "D"      # 日线
    WEEKLY = "W"     # 周线


@dataclass
class Bar:
    """K线数据"""
    symbol: str                # 股票代码
    datetime: datetime         # 时间
    frequency: Frequency       # 频率
    
    # 基础OHLC数据
    open: float               # 开盘价
    high: float               # 最高价
    low: float                # 最低价
    close: float              # 收盘价
    volume: int               # 成交量
    amount: float             # 成交额
    turnover: float = 0.0     # 换手率
    
    # 技术指标
    ma5: Optional[float] = None              # 5日均线
    ma20: Optional[float] = None             # 20日均线
    ma60: Optional[float] = None             # 60日均线
    macd_dif: Optional[float] = None         # MACD DIF线
    macd_dea: Optional[float] = None         # MACD DEA线
    macd_histogram: Optional[float] = None   # MACD柱状图
    rsi_14: Optional[float] = None           # 14日RSI(0-100)
    boll_upper: Optional[float] = None       # 布林带上轨
    boll_lower: Optional[float] = None       # 布林带下轨
    
    # 基本面数据
    market_cap: Optional[float] = None               # 市值
    circulating_market_cap: Optional[float] = None  # 流通市值
    is_st: bool = False                              # 是否ST
    is_new_stock: bool = False                       # 是否新股次新股
    
    def __post_init__(self):
        """验证K线数据的有效性"""
        if self.high < max(self.open, self.close):
            raise ValueError("最高价不能低于开盘价或收盘价")
        if self.low > min(self.open, self.close):
            raise ValueError("最低价不能高于开盘价或收盘价")
        if self.volume < 0:
            raise ValueError("成交量不能为负数")
        if self.amount < 0:
            raise ValueError("成交额不能为负数")
    
    @property
    def is_bullish(self) -> bool:
        """是否为阳线"""
        return self.close > self.open
    
    @property
    def is_bearish(self) -> bool:
        """是否为阴线"""
        return self.close < self.open
    
    @property
    def body_size(self) -> float:
        """实体大小（绝对值）"""
        return abs(self.close - self.open)
    
    @property
    def body_size_pct(self) -> float:
        """实体大小百分比"""
        return self.body_size / self.open if self.open > 0 else 0
    
    @property
    def upper_shadow(self) -> float:
        """上影线长度"""
        return self.high - max(self.open, self.close)
    
    @property
    def lower_shadow(self) -> float:
        """下影线长度"""
        return min(self.open, self.close) - self.low
    
    @property
    def change_pct(self) -> float:
        """涨跌幅"""
        return (self.close - self.open) / self.open if self.open > 0 else 0
    
    @property
    def has_technical_indicators(self) -> bool:
        """是否包含技术指标"""
        return any([
            self.ma5 is not None,
            self.ma20 is not None,
            self.macd_dif is not None,
            self.rsi_14 is not None
        ])
    
    def __str__(self) -> str:
        """字符串表示"""
        return (f"Bar({self.symbol} {self.datetime} "
                f"OHLC={self.open:.2f}/{self.high:.2f}/{self.low:.2f}/{self.close:.2f} "
                f"Vol={self.volume})")